src/components/MultiSelect/MultiSelect.tsx

```diff
@@ -1,138 +1,435 @@
+
+        const filteredItems = items.filter((item) => {
+          if (item.children) {
+            if (item.name.toLowerCase().includes(value)) {
+              return true;
+            }
+
+            return item.children.some((childItem) =>
+              childItem.name.toLowerCase().includes(value)
+            );
+          }
+
+          return item.name.toLowerCase().includes(value);
+        });
+
+        setItemsList(filteredItems);
+      };
+
+      // The clearSearchKeyword method resets the search keyword in the MultiSelect component.
+      // If the listOverflow is set to "scroll," it sets the list of items to the full list (items).
+      // If listOverflow is set to a value other than "scroll,"
+      // it checks if the current count of displayed items (listItemsCount) is equal to the default number of visible items (defaultItemsVisible).
+      // If true, it displays the default items; otherwise, it sets the list of items to the full list.
+      // This method is responsible for clearing the search and displaying the appropriate items based on the listOverflow and item count conditions.
+      const clearSearchKeyword = () => {
+        if (listOverflow === "scroll") {
+          setItemsList(items);
+        } else {
+          listItemsCount === defaultItemsVisible
+            ? displayDefaultItems()
+            : setItemsList(items);
+        }
+      };
+
+      const displayDefaultItems = useCallback(() => {
+        const list = [];
+        let count = 0;
+        for (let i = 0; i < items.length && count < defaultItemsVisible; i++) {
+          const currentItem = items[i];
+          count++;
+          list.push(currentItem);
+          if (count >= defaultItemsVisible) {
+            break;
+          }
+        }
+        setItemsList(list);
+      }, [items, defaultItemsVisible, setItemsList]);
+
+      // Separate effect for handling listOverflow "expand"
+      React.useEffect(() => {
+        if (listOverflow === "expand") {
+          setListHeight("");
+          if (listItemsCount === defaultItemsVisible) {
+            displayDefaultItems();
+          } else {
+            setItemsList(items);
+            setViewAllLabel("View less");
+          }
+        }
+      }, [
+        listOverflow,
+        listItemsCount,
+        items,
+        defaultItemsVisible,
+        displayDefaultItems,
+      ]);
+
+      const showSearchInputBox = () => (
+        <>
+          {isSearchable && (
+            <TextInput
+              id="multi-select-text-input-id"
+              labelText={`Search ${buttonText}`}
+              isClearable={true}
+              isClearableCallback={clearSearchKeyword}
+              placeholder={`Search ${buttonText}`}
+              onChange={onChangeSearch}
+              showLabel={false}
+              showRequiredLabel={false}
+              type="text"
+              __css={styles.menuSearchInputBox}
+              marginBottom="s"
+            />
+          )}
```

Comment:
If this is the only thing being returned from this function, then you don't need the react fragment `<>` wrapper.